File: scripts/seed.py
#!/usr/bin/env python3
"""Load contracts/fixtures/ into a fresh SQLite database.

The seeded DB is not decorative: it is calibrated so that the API reproduces the
fixtures byte-for-byte. `GET /spaces/redesign/canvas` returns canvas.json,
`?include_deleted=true` returns canvas.with-deleted.json, and
`GET /spaces/redesign/feedback?actor=claude-code` — with no `since` — returns
feedback.claude-code.since-12.json, because claude-code's stored cursor is seeded
at 12. tests/contract/test_fixtures_roundtrip.py asserts exactly that.

Usage:
    python scripts/seed.py                    # seed the default DB, refuse if it exists
    python scripts/seed.py --reset            # delete the DB first
    python scripts/seed.py --db /tmp/x.db --media-dir /tmp/media
"""

from __future__ import annotations

import argparse
import json
import sqlite3
import struct
import sys
import zlib
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO_ROOT))

from server import config  # noqa: E402

FIXTURES = REPO_ROOT / "contracts" / "fixtures"


def load(name: str):
    return json.loads((FIXTURES / name).read_text())


def apply_schema(conn: sqlite3.Connection) -> None:
    conn.executescript(config.SCHEMA_PATH.read_text())


def last_ts(events: list[dict], subject_id: str, default: str) -> str:
    """updated_at is not in the fixtures; derive it from the event log."""
    stamps = [e["ts"] for e in events if e["subject_id"] == subject_id]
    return stamps[-1] if stamps else default


def seed(conn: sqlite3.Connection, media_root: Path) -> str:
    space = load("space.json")
    canvas = load("canvas.with-deleted.json")
    annotations = load("annotations.json")
    events = load("events.json")["events"]

    space_id = space["id"]
    created_at = space["created_at"]

    conn.execute(
        "INSERT INTO space (id, slug, title, revision_mode, seq, created_at)"
        " VALUES (?, ?, ?, ?, ?, ?)",
        (space_id, space["slug"], space["title"], space["revision_mode"],
         space["seq"], created_at),
    )

    # Cards. Rows are inserted in fixture order so `ORDER BY rowid` reproduces it.
    # sp_deleted_at is a read-time projection of card.deleted_at, not stored in the
    # node blob — otherwise GET /canvas (live only) would leak it.
    for node in canvas["nodes"]:
        node = dict(node)
        deleted_at = node.pop("sp_deleted_at", None)
        conn.execute(
            "INSERT INTO card (id, space_id, node_json, rev, created_by, updated_at,"
            " deleted_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
            (
                node["id"], space_id, json.dumps(node, ensure_ascii=False),
                node.get("sp_rev", 1), node.get("sp_created_by", "claude-code"),
                last_ts(events, node["id"], created_at), deleted_at,
            ),
        )

    for edge in canvas["edges"]:
        conn.execute(
            "INSERT INTO link (id, space_id, edge_json, created_by, updated_at,"
            " deleted_at) VALUES (?, ?, ?, ?, ?, NULL)",
            (
                edge["id"], space_id, json.dumps(edge, ensure_ascii=False),
                edge.get("sp_created_by", "claude-code"),
                last_ts(events, edge["id"], created_at),
            ),
        )

    resolved_ts = {
        e["subject_id"]: e["ts"] for e in events if e["type"] == "annotation.resolved"
    }
    for ann in annotations:
        # card_title and stale are computed on read; never stored.
        conn.execute(
            "INSERT INTO annotation (id, space_id, card_id, card_rev, selector, body,"
            " motivation, creator, creator_kind, resolved, resolved_reply, resolved_at,"
            " created_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)",
            (
                ann["id"], space_id, ann["card_id"], ann["card_rev"],
                json.dumps(ann["selector"]) if ann["selector"] is not None else None,
                ann["body"], ann["motivation"], ann["creator"], ann["creator_kind"],
                1 if ann["resolved"] else 0, ann.get("resolved_reply"),
                resolved_ts.get(ann["id"]) if ann["resolved"] else None,
                ann["created_at"],
            ),
        )

    for ev in events:
        conn.execute(
            "INSERT INTO event (space_id, seq, ts, type, subject_id, actor, actor_kind,"
            " payload) VALUES (?,?,?,?,?,?,?,?)",
            (
                space_id, ev["seq"], ev["ts"], ev["type"], ev["subject_id"],
                ev["actor"], ev["actor_kind"],
                json.dumps(ev["payload"]) if ev.get("payload") is not None else None,
            ),
        )

    # The cursor that makes the default feedback call reproduce the fixture.
    conn.execute(
        "INSERT INTO actor_cursor (space_id, actor, seq) VALUES (?, ?, ?)",
        (space_id, "claude-code", 12),
    )

    conn.commit()
    write_fixture_media(media_root / space_id)
    return space_id


def write_fixture_media(dest: Path) -> None:
    """c_shot is a file node pointing at .../media/m_01.png. Give it a real file."""
    dest.mkdir(parents=True, exist_ok=True)
    (dest / "m_01.png").write_bytes(_placeholder_png(360, 280))


def _placeholder_png(width: int, height: int) -> bytes:
    """A valid greyscale PNG, no dependencies. Diagonal stripes so it is obviously
    a placeholder and not a broken image."""

    def chunk(tag: bytes, data: bytes) -> bytes:
        return (struct.pack(">I", len(data)) + tag + data
                + struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF))

    raw = bytearray()
    for y in range(height):
        raw.append(0)  # filter: none
        for x in range(width):
            raw.append(0xE8 if ((x + y) // 12) % 2 else 0xC8)
    return (
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 0, 0, 0, 0))
        + chunk(b"IDAT", zlib.compress(bytes(raw), 9))
        + chunk(b"IEND", b"")
    )


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--db", type=Path, default=None, help="defaults to ANALOG_DB / data/analog.db")
    ap.add_argument("--media-dir", type=Path, default=None, help="defaults to data/media")
    ap.add_argument("--reset", action="store_true", help="delete an existing DB first")
    args = ap.parse_args(argv)

    db = args.db or config.db_path()
    media = args.media_dir or config.media_dir()
    db.parent.mkdir(parents=True, exist_ok=True)

    if db.exists():
        if not args.reset:
            print(f"{db} already exists; pass --reset to replace it", file=sys.stderr)
            return 1
        for suffix in ("", "-wal", "-shm"):
            Path(str(db) + suffix).unlink(missing_ok=True)

    conn = sqlite3.connect(db)
    try:
        apply_schema(conn)
        space_id = seed(conn, media)
    finally:
        conn.close()

    print(f"seeded {db}")
    print(f"  space   redesign ({space_id}) — 6 live cards, 1 deleted, 4 links")
    print(f"  cursor  claude-code @ 12")
    print(f"  media   {media / space_id / 'm_01.png'}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: cmd/analog-server/seed.go
package main

import (
	"bytes"
	"compress/zlib"
	"database/sql"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"hash/crc32"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"

	"github.com/meowkey-dev/analog/internal/config"
	"github.com/meowkey-dev/analog/internal/store"
)

// The seeded DB is not decorative: it is calibrated so that the API reproduces the
// fixtures byte-for-byte. GET /spaces/redesign/canvas returns canvas.json,
// ?include_deleted=true returns canvas.with-deleted.json, and
// GET /spaces/redesign/feedback?actor=claude-code — with no `since` — returns
// feedback.claude-code.since-12.json, because claude-code's stored cursor is seeded
// at 12. tests/contract/test_fixtures_roundtrip.py asserts exactly that.

func seedCmd() *cobra.Command {
	var dbPath, mediaDir, fixtures string
	var reset bool

	cmd := &cobra.Command{
		Use:   "seed",
		Short: "Load contracts/fixtures/ into a fresh database",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			if dbPath == "" {
				dbPath = config.DBPath()
			}
			if mediaDir == "" {
				mediaDir = config.MediaDir()
			}
			return runSeed(dbPath, mediaDir, fixtures, reset)
		},
	}
	cmd.Flags().StringVar(&dbPath, "db", "", "defaults to ANALOG_DB / data/analog.db")
	cmd.Flags().StringVar(&mediaDir, "media-dir", "", "defaults to data/media")
	cmd.Flags().StringVar(&fixtures, "fixtures", filepath.Join("contracts", "fixtures"),
		"directory holding the frozen fixtures")
	cmd.Flags().BoolVar(&reset, "reset", false, "delete an existing DB first")
	return cmd
}

// jsonObject and jsonArray keep numeric literals exactly as the fixtures wrote them.
func loadFixture(dir, name string, into any) error {
	raw, err := os.ReadFile(filepath.Join(dir, name))
	if err != nil {
		return fmt.Errorf("reading %s: %w", name, err)
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	return dec.Decode(into)
}

type fixtureEvent struct {
	Seq       int64          `json:"seq"`
	TS        string         `json:"ts"`
	Type      string         `json:"type"`
	SubjectID string         `json:"subject_id"`
	Actor     string         `json:"actor"`
	ActorKind string         `json:"actor_kind"`
	Payload   map[string]any `json:"payload"`
}

func runSeed(dbPath, mediaDir, fixtures string, reset bool) error {
	if _, err := os.Stat(dbPath); err == nil {
		if !reset {
			fmt.Fprintf(os.Stderr, "%s already exists; pass --reset to replace it\n", dbPath)
			os.Exit(1)
		}
		for _, suffix := range []string{"", "-wal", "-shm"} {
			_ = os.Remove(dbPath + suffix)
		}
	}
	if err := os.MkdirAll(filepath.Dir(dbPath), 0o755); err != nil {
		return err
	}

	var space map[string]any
	var canvas struct {
		Nodes []map[string]any `json:"nodes"`
		Edges []map[string]any `json:"edges"`
	}
	var annotations []map[string]any
	var eventFile struct {
		Events []fixtureEvent `json:"events"`
	}
	for _, load := range []func() error{
		func() error { return loadFixture(fixtures, "space.json", &space) },
		func() error { return loadFixture(fixtures, "canvas.with-deleted.json", &canvas) },
		func() error { return loadFixture(fixtures, "annotations.json", &annotations) },
		func() error { return loadFixture(fixtures, "events.json", &eventFile) },
	} {
		if err := load(); err != nil {
			return err
		}
	}

	db, err := sql.Open("sqlite", "file:"+dbPath+"?_pragma=foreign_keys(ON)")
	if err != nil {
		return err
	}
	defer db.Close()
	if _, err := db.Exec(store.Schema()); err != nil {
		return err
	}

	spaceID := str(space["id"])
	createdAt := str(space["created_at"])

	// updated_at is not in the fixtures; derive it from the event log.
	lastTS := func(subjectID string) string {
		out := createdAt
		for _, e := range eventFile.Events {
			if e.SubjectID == subjectID {
				out = e.TS
			}
		}
		return out
	}

	if _, err := db.Exec(
		"INSERT INTO space (id, slug, title, revision_mode, seq, created_at)"+
			" VALUES (?,?,?,?,?,?)",
		spaceID, str(space["slug"]), str(space["title"]), str(space["revision_mode"]),
		num(space["seq"]), createdAt); err != nil {
		return err
	}

	// Rows are inserted in fixture order so `ORDER BY rowid` reproduces it.
	// sp_deleted_at is a read-time projection of card.deleted_at, not stored in the
	// node blob — otherwise GET /canvas (live only) would leak it.
	for _, raw := range canvas.Nodes {
		node := map[string]any{}
		for k, v := range raw {
			if k != "sp_deleted_at" {
				node[k] = v
			}
		}
		var deletedAt any
		if v, ok := raw["sp_deleted_at"]; ok {
			deletedAt = v
		}
		rev := int64(1)
		if v, ok := node["sp_rev"]; ok {
			rev = num(v)
		}
		createdBy := "claude-code"
		if v := str(node["sp_created_by"]); v != "" {
			createdBy = v
		}
		blob, err := json.Marshal(node)
		if err != nil {
			return err
		}
		if _, err := db.Exec(
			"INSERT INTO card (id, space_id, node_json, rev, created_by, updated_at,"+
				" deleted_at) VALUES (?,?,?,?,?,?,?)",
			str(node["id"]), spaceID, string(blob), rev, createdBy,
			lastTS(str(node["id"])), deletedAt); err != nil {
			return err
		}
	}

	for _, edge := range canvas.Edges {
		blob, err := json.Marshal(edge)
		if err != nil {
			return err
		}
		createdBy := "claude-code"
		if v := str(edge["sp_created_by"]); v != "" {
			createdBy = v
		}
		if _, err := db.Exec(
			"INSERT INTO link (id, space_id, edge_json, created_by, updated_at,"+
				" deleted_at) VALUES (?,?,?,?,?,NULL)",
			str(edge["id"]), spaceID, string(blob), createdBy,
			lastTS(str(edge["id"]))); err != nil {
			return err
		}
	}

	resolvedTS := map[string]string{}
	for _, e := range eventFile.Events {
		if e.Type == "annotation.resolved" {
			resolvedTS[e.SubjectID] = e.TS
		}
	}
	for _, ann := range annotations {
		// card_title and stale are computed on read; never stored.
		var selector any
		if v, ok := ann["selector"]; ok && v != nil {
			blob, err := json.Marshal(v)
			if err != nil {
				return err
			}
			selector = string(blob)
		}
		resolved := 0
		var resolvedAt any
		if b, ok := ann["resolved"].(bool); ok && b {
			resolved = 1
			if ts, ok := resolvedTS[str(ann["id"])]; ok {
				resolvedAt = ts
			}
		}
		var reply any
		if v, ok := ann["resolved_reply"]; ok && v != nil {
			reply = v
		}
		if _, err := db.Exec(
			"INSERT INTO annotation (id, space_id, card_id, card_rev, selector, body,"+
				" motivation, creator, creator_kind, resolved, resolved_reply, resolved_at,"+
				" created_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)",
			str(ann["id"]), spaceID, str(ann["card_id"]), num(ann["card_rev"]),
			selector, str(ann["body"]), str(ann["motivation"]), str(ann["creator"]),
			str(ann["creator_kind"]), resolved, reply, resolvedAt,
			str(ann["created_at"])); err != nil {
			return err
		}
	}

	for _, e := range eventFile.Events {
		var payload any
		if e.Payload != nil {
			blob, err := json.Marshal(e.Payload)
			if err != nil {
				return err
			}
			payload = string(blob)
		}
		if _, err := db.Exec(
			"INSERT INTO event (space_id, seq, ts, type, subject_id, actor, actor_kind,"+
				" payload) VALUES (?,?,?,?,?,?,?,?)",
			spaceID, e.Seq, e.TS, e.Type, e.SubjectID, e.Actor, e.ActorKind,
			payload); err != nil {
			return err
		}
	}

	// The cursor that makes the default feedback call reproduce the fixture.
	if _, err := db.Exec(
		"INSERT INTO actor_cursor (space_id, actor, seq) VALUES (?,?,?)",
		spaceID, "claude-code", 12); err != nil {
		return err
	}

	// c_shot is a file node pointing at .../media/m_01.png. Give it a real file.
	mediaTarget := filepath.Join(mediaDir, spaceID)
	if err := os.MkdirAll(mediaTarget, 0o755); err != nil {
		return err
	}
	if err := os.WriteFile(filepath.Join(mediaTarget, "m_01.png"),
		placeholderPNG(360, 280), 0o644); err != nil {
		return err
	}

	fmt.Printf("seeded %s\n", dbPath)
	fmt.Printf("  space   redesign (%s) — 6 live cards, 1 deleted, 4 links\n", spaceID)
	fmt.Printf("  cursor  claude-code @ 12\n")
	fmt.Printf("  media   %s\n", filepath.Join(mediaTarget, "m_01.png"))
	return nil
}

func str(v any) string {
	s, _ := v.(string)
	return s
}

func num(v any) int64 {
	if n, ok := v.(json.Number); ok {
		i, _ := n.Int64()
		return i
	}
	if f, ok := v.(float64); ok {
		return int64(f)
	}
	return 0
}

// placeholderPNG is a valid greyscale PNG with diagonal stripes, so it is obviously
// a placeholder and not a broken image.
func placeholderPNG(width, height int) []byte {
	chunk := func(tag string, data []byte) []byte {
		var out bytes.Buffer
		_ = binary.Write(&out, binary.BigEndian, uint32(len(data)))
		body := append([]byte(tag), data...)
		out.Write(body)
		_ = binary.Write(&out, binary.BigEndian, crc32.ChecksumIEEE(body))
		return out.Bytes()
	}

	raw := make([]byte, 0, height*(width+1))
	for y := 0; y < height; y++ {
		raw = append(raw, 0) // filter: none
		for x := 0; x < width; x++ {
			if ((x+y)/12)%2 != 0 {
				raw = append(raw, 0xE8)
			} else {
				raw = append(raw, 0xC8)
			}
		}
	}
	var compressed bytes.Buffer
	zw, _ := zlib.NewWriterLevel(&compressed, zlib.BestCompression)
	_, _ = zw.Write(raw)
	_ = zw.Close()

	var ihdr bytes.Buffer
	_ = binary.Write(&ihdr, binary.BigEndian, uint32(width))
	_ = binary.Write(&ihdr, binary.BigEndian, uint32(height))
	ihdr.Write([]byte{8, 0, 0, 0, 0}) // 8-bit greyscale, no interlace

	var out bytes.Buffer
	out.Write([]byte("\x89PNG\r\n\x1a\n"))
	out.Write(chunk("IHDR", ihdr.Bytes()))
	out.Write(chunk("IDAT", compressed.Bytes()))
	out.Write(chunk("IEND", nil))
	return out.Bytes()
}
